packages/itwin/grouping-mapping-widget/src/widget/components/PropertyMenu.tsx

```diff
@@ -2,404 +2,212 @@
   );
-  const { isLoading: isLoadingGroupProperties, data: groupProperties, refreshData: refreshGroupProperties } =
-    useCombinedFetchRefresh<IGroupPropertyTyped>(fetchGroupProperties);
+  const { isLoading: isLoadingGroupProperties, refreshData: refreshGroupProperties } =
+    useCombinedFetchRefresh<GroupProperty>(fetchGroupProperties, setGroupProperties);
 
   const fetchCalculatedProperties = useMemo(
     () => {
       return async () => mappingClient.getCalculatedProperties((await getAccessToken()), iModelId, mappingId, groupId);
     },
     [getAccessToken, mappingClient, iModelId, mappingId, groupId],
   );
-  const { isLoading: isLoadingCalculatedProperties, data: calculatedProperties, refreshData: refreshCalculatedProperties } =
-    useCombinedFetchRefresh<ICalculatedPropertyTyped>(fetchCalculatedProperties);
+  const { isLoading: isLoadingCalculatedProperties, refreshData: refreshCalculatedProperties } =
+    useCombinedFetchRefresh<CalculatedProperty>(fetchCalculatedProperties, setCalculatedProperties);
 
   const fetchCustomCalculations = useMemo(
     () => {
       return async () => mappingClient.getCustomCalculations((await getAccessToken()), iModelId, mappingId, groupId);
     },
     [getAccessToken, mappingClient, iModelId, mappingId, groupId],
   );
-  const { isLoading: isLoadingCustomCalculations, data: customCalculations, refreshData: refreshCustomCalculations } =
-    useCombinedFetchRefresh<ICustomCalculationTyped>(fetchCustomCalculations);
+  const { isLoading: isLoadingCustomCalculations, refreshData: refreshCustomCalculations } =
+    useCombinedFetchRefresh<CustomCalculation>(fetchCustomCalculations, setCustomCalculationProperties);
 
   useEffect(() => {
     const initialize = async () => {
-      try {
-        const query = group.query;
-        const queryRowCount = await iModelConnection.queryRowCount(query);
-        if (queryRowCount === 0) {
-          toaster.warning("The query is valid but produced no results.");
-          await goBack();
-        }
-        const keys = await manufactureKeys(query, iModelConnection);
-        setKeySet(keys);
+      if (!iModelConnection) return;
+      setIsLoading(true);
+      clearEmphasizedOverriddenElements();
+      if (showGroupColor) {
+        const result = await getHiliteIdsAndKeysetFromGroup(iModelConnection, group, hilitedElementsQueryCache);
         Presentation.selection.clearSelection(
           "GroupingMappingWidget",
           iModelConnection,
         );
-        clearEmphasizedOverriddenElements();
-        const resolvedIds = await visualizeElementsByKeys(keys, "red");
-        await zoomToElements(resolvedIds);
-        setResolvedHiliteIds(resolvedIds);
-        setIsLoading(false);
-      } catch {
-        toaster.negative(`Could not load ${group.groupName}.`);
-        await goBack();
+        visualizeElements(result.ids, color);
+        await zoomToElements(result.ids);
       }
+      setIsLoading(false);
```

Comment:
If error occurs in any of the above async operations the widget will be stuck at loading state.